### ISSUE TITLE
Not closing or aborting communication object on exit

### ISSUE THREAD

**Issue by madsbk** [dask/distributed#3186]
While debugging the known [`LocalCluster(processes=True,...)`](https://github.com/dask/distributed/blob/9429ffe735e3767b427fd0b18aaa29d6a7da8513/distributed/comm/tests/test_ucx.py#L265) bug, I found out that `UCX.close()` or `UCX.abort()` is never called by the worker process when exiting. After the worker terminates, the schduler tries to send a message to the worker, which fails.

What is the exact scheduler-worker protocol? Is it by design that the scheduler tries to send messages to a worker it has killed?

cc. @quasiben @mrocklin @pentschev

**Comment by mrocklin** [comment]
I just tried this with tcp, and did see that we at least call the close and abort in that case.

```diff
diff --git a/distributed/comm/tcp.py b/distributed/comm/tcp.py
index d0322e15..72c933f8 100644
--- a/distributed/comm/tcp.py
+++ b/distributed/comm/tcp.py
@@ -262,6 +262,7 @@ class TCP(Comm):

     @gen.coroutine
     def close(self):
+        print("close!")
         # We use gen.coroutine here rather than async def to avoid errors like
         # Task was destroyed but it is pending!
         # Triggered by distributed.deploy.tests.test_local::test_silent_startup
@@ -279,6 +280,7 @@ class TCP(Comm):
                 stream.close()

     def abort(self):
+        print("abort!")
         stream, self.stream = self.stream, None
         if stream is not None and not stream.closed():
             self._finalizer.detach()
```

```python
In [1]: from dask.distributed import LocalCluster

In [2]: cluster = LocalCluster(processes=True)

In [3]: cluster.close()
abort!
abort!
abort!
abort!
close!
close!
close!
close!
abort!
close!
close!
close!

close!
In [4]:
Do you really want to exit ([y]/n)? ^D
(dev) ngvpn01-160-176:distributed mrocklin$ ipython
fPython 3.7.4 (default, Aug 13 2019, 15:17:50)
Type 'copyright', 'credits' or 'license' for more information
IPython 7.7.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: from dask.distributed import LocalCluster

In [2]: cluster = LocalCluster(processes=True)

In [3]:
Do you really want to exit ([y]/n)? ^D
close!
close!
close!
close!
abort!
close!
close!
close!
close!
(dev) ngvpn01-160-176:distributed mrocklin$ diff
```

> What is the exact scheduler-worker protocol?

I'm not sure exactly what level of detail you're looking for here, but here is a rough outline of what happens.

1.  Worker comes alive, and tells scheduler that it exists
2.  Scheduler opens a new long-running connection to the worker
3.  When the worker closes down it sends a message to the scheduler unregistering itself
4.  The scheduler closes down all of its connections to that worker
5.  The worker closes down its connection pool

> Is it by design that the scheduler tries to send messages to a worker it has killed?

Not to my knowledge.

**Comment by madsbk** [comment]
The scheduler tries to send messages to killed workers even when  `LocalCluster` uses TCP. 
When running `test_ucx.py::test_ucx_localcluster[True]` with
```python
with LocalCluster(protocol="tcp", ...)
```
I get:
```
TCP COMM create - pid: 12270, count: 2
PROCESS START pid: 12299                <== New process pid: 12299
TCP COMM create - pid: 12299, count: 1  <== Creating class UCX(Comm)
TCP COMM create - pid: 12270, count: 3
PROCESS START pid: 12300
TCP COMM create - pid: 12300, count: 1
TCP COMM create - pid: 12270, count: 4
TCP COMM create - pid: 12270, count: 5
TCP COMM create - pid: 12270, count: 6
TCP COMM create - pid: 12270, count: 7
TCP COMM create - pid: 12270, count: 8
TCP COMM create - pid: 12270, count: 9
TCP COMM create - pid: 12300, count: 2
TCP COMM close - pid: 12270, count: 1
TCP COMM abort - pid: 12270, count: 2
TCP COMM abort - pid: 12270, count: 3
TCP COMM create - pid: 12270, count: 10
TCP COMM create - pid: 12270, count: 11
PROCESS EXIT pid: 12299                 <== Exit process pid: 12299
TCP COMM close - pid: 12300, count: 1
TCP COMM close - pid: 12270, count: 4
TCP COMM close - pid: 12300, count: 2
TCP COMM close - pid: 12270, count: 5
PROCESS EXIT pid: 12300
TCP COMM close - pid: 12270, count: 6
TCP COMM abort - pid: 12270, count: 7
TCP COMM close - pid: 12270, count: 8
TCP COMM close - pid: 12270, count: 9
TCP COMM close - pid: 12270, count: 10
TCP COMM close - pid: 12270, count: 11
PASSED
```
Notice, process `12299` exists before calling close/abort.
With #3178 and https://github.com/rapidsai/ucx-py/pull/309, UCX support this design.
